app/src/main/java/tech/ula/model/state/SessionStartupFsm.kt

```diff
@@ -164,38 +166,28 @@ class SessionStartupFsm(
     }
 
     private fun handleDownloadAssets(assetsToDownload: List<Asset>) {
-        downloadingIds.clear()
-        downloadedIds.clear()
-
         // If the state isn't updated first, AssetDownloadComplete events will be submitted before
         // the transition is acceptable.
-        state.postValue(DownloadingRequirements(0, assetsToDownload.size))
-        val newDownloads = downloadUtility.downloadRequirements(assetsToDownload)
-        downloadingIds.addAll(newDownloads)
+        state.postValue(DownloadingAssets(0, assetsToDownload.size))
+        downloadUtility.downloadRequirements(assetsToDownload)
     }
 
     private fun handleAssetsDownloadComplete(downloadId: Long) {
-        if (!downloadUtility.downloadedSuccessfully(downloadId)) {
-            val reason = downloadUtility.getReasonForDownloadFailure(downloadId)
-            state.postValue(DownloadsHaveFailed(reason))
-            return
-        }
-
-        downloadedIds.add(downloadId)
-        downloadUtility.setTimestampForDownloadedFile(downloadId)
-        if (downloadingIds.size != downloadedIds.size) {
-            state.postValue(DownloadingRequirements(downloadedIds.size, downloadingIds.size))
-            return
-        }
+        val result = downloadUtility.handleDownloadComplete(downloadId)
+        handleAssetDownloadState(result)
+    }
 
-        downloadedIds.sort()
-        downloadingIds.sort()
-        if (downloadedIds != downloadingIds) {
-            state.postValue(DownloadsHaveFailed("Downloads completed with non-enqueued downloads"))
-            return
+    private fun handleAssetDownloadState(assetDownloadState: AssetDownloadState) {
+        return when (assetDownloadState) {
+            // We don't care if we've something else has downloaded something.
```

Comment:
Looks like there's an extra word somewhere in this line

Reply:
I grammar goodly